### PR TITLE
[libc++] Deprecates std::errc constants.

### PR DESCRIPTION
**Pull request by mordante**
Implements:
- LWG3869 Deprecate std::errc constants related to UNIX STREAMS

### CODE REVIEW

File: libcxx/include/__system_error/errc.h [hunk 1]
```diff
@@ -113,6 +113,15 @@ _LIBCPP_BEGIN_NAMESPACE_STD
...
 
 // enum class errc
+//
+// LWG3869 deprecates the UNIX STREAMS macros and enum values.
+// This makes the code clumbersome:
+// - the enum value is deprecated and should show a diagnostic,
+// - the macro is deprecated and should _not_ show a diagnostic in this
+//   context, and
+// - the macro is not always available.
+// This leads to the odd pushing and popping of the deprecated
```
Comment (philnik777) [inline libcxx/include/__system_error/errc.h:123]: If a declaration itself is deprecated and uses deprecated functionality, the deprecation warning of the stuff it uses should be suppressed: https://godbolt.org/z/b839bqGj7. This looks like a Clang bug to me. Could you file one?

Comment (mordante) [reply libcxx/include/__system_error/errc.h:123]: @AaronBallman do you agree it's a bug? If so I'm happy to file a report.

Comment (AaronBallman) [reply libcxx/include/__system_error/errc.h:123]: Oofda. I want to call this a bug, but it's going to be exceptionally hard to work around because of the difference in phases of translation. We issue the deprecated macro warning when the macro is expanded, which is *long* before we have any idea of how the value is being used. So it may be worth filing a QoI enhancement request over it, but I don't know that we have a way to improve it just yet.

Comment (mordante) [reply libcxx/include/__system_error/errc.h:123]: Thanks I've filed https://github.com/llvm/llvm-project/issues/81359

File: libcxx/src/random.cpp [hunk 2]
```diff
@@ -79,8 +79,10 @@ unsigned random_device::operator()() {
   char* p  = reinterpret_cast<char*>(&r);
   while (n > 0) {
     ssize_t s = read(__f_, p, n);
+    _LIBCPP_SUPPRESS_DEPRECATED_PUSH
     if (s == 0)
-      __throw_system_error(ENODATA, "random_device got EOF");
+      __throw_system_error(ENODATA, "random_device got EOF"); // TODO ENODATA -> ENOMSG
```
Comment (mordante) [inline libcxx/src/random.cpp:84]: I like to do this is a separate commit since the result is observable. (https://github.com/llvm/llvm-project/issues/81360)

File: libcxx/include/__system_error/errc.h [hunk 3]
```diff
@@ -58,18 +58,18 @@ enum class errc
     no_child_process,                   // ECHILD
     no_link,                            // ENOLINK
     no_lock_available,                  // ENOLCK
-    no_message_available,               // ENODATA
+    no_message_available,               // ENODATA         // deprecated
```
Comment (ldionne) [inline libcxx/include/__system_error/errc.h:61]: I think you're missing an update to the LWG issue status.

**Review by ldionne — APPROVED** [review]

**Comment by nico** [comment]
We get diags about `ENOSTR` now. Is that intentional? (I'm asking because the commit description talks about std::errc, but `ENOSTR` is not `std::errc::not_a_stream`.)

…hm, looking at libcxx/include/cerrno, it looks pretty intentional. But the commit description and the changelog entry could maybe call this out more (?).

**Comment by nico** [comment]
Is there a way to opt out of this, to be able to fix async? I thought maybe defining `_LIBCPP_SUPPRESS_DEPRECATED_PUSH` might work, but that apparently isn't an extension point.

**Comment by mordante** [comment]
Fair point the message could have been clearer. What do you mean with an opt out? This is a deprecation that we like people to fix, I expect there will be a removal which will be retroactively applied. (Note this is a change by POSIX and C++ follows its lead.)

**Comment by nico** [comment]
I mean some way (preprocessor define I suppose) that says "I have seen this deprecation message, please silence it for now". That's useful to be able to update your libc++ and then work on new deprecations asynchronously. Without opt-outs, you always have to fix all issues before you can updating again. That opens the risk that upstream lands some bug that you don't notice until the next update attempt, and if that adds a new deprecation before the new bug, you now have to fix _that_ deprecation before you can roll in the bug fix.

MSVC's STL has opt-out macros for every deprecation they add, as far as I know. That is very useful for consumers, since it makes updating independent from fixing deprecation warnings. (It's enough to have these opt-outs for a little while, not forever, of course.)

(For some data points, https://autoroll.skia.org/r/libcxx-chromium/roll-history shows that libc++ updates have been very rough for us this year, with several weeks between successful updates. When things work, we usually update within hours. The last time, it took us a few weeks to clean up the iterator change in #74482 since that didn't have an effective opt-out. In the meantime, 79 changes landed. Several of them cleaned up includes, but luckily it looks like we only had to add a single `#include` line for that. And then this here landed. Which might not need a _ton_ of cleanup, not sure yet, but it's in a dependency, not in direct code, so it'll take a while to pick up the fix again. Every single of these changes isn't super big, but there have been like 5 medium-sized once this year so far, and it's just so much easier if there's an opt-out flag.)

**Comment by mordante** [comment]
Thanks for the additional information.

Typically we only add opt-out macros when we remove features, not when we add deprecation flags. I think usually these depredations only affect newer language versions. But I've added some for long deprecated features too; they were added before we had deprecated messages.

I've no objection to add an opt-out macro. @ldionne any objection from your side? And should we always do this when we add a new deprecation? For example. for barrier, latch, and friends prior to C++20.

**Comment by ldionne** [comment]
We provide `_LIBCPP_DISABLE_DEPRECATION_WARNINGS` to allow turning those off. Sure, that will turn off *all* deprecation warnings, but in practice is it really a problem if it takes only a day or so to clean up the deprecated uses in most cases?

Alternatively, we could systematically introduce such macros but this is stuff that we'd be stuck with "forever" unless we agree on some kind of process to clean them up after N releases. If we do that, I'd argue that we want to remove them after a single release: if you need to silence deprecation warnings for an extended period of time, I would argue you shouldn't compile with deprecation warnings turned into errors.

> I've no objection to add an opt-out macro. @ldionne any objection from your side? And should we always do this when we add a new deprecation? For example. for barrier, latch, and friends prior to C++20.

This is an example where the deprecation + removal process is made more complicated. So we would do this, I guess?
- LLVM 19: Deprecation warning for `<barrier>` & friends in < C++20, with `_LIBCPP_DISABLE_DEPRECATION_WARNING_SYNC_BEFORE_CXX20`
- LLVM 20: Remove `_LIBCPP_DISABLE_DEPRECATION_WARNING_SYNC_BEFORE_CXX20`, at the same time remove `<barrier>` & friends from < C++20 ?

Or would the deprecation-warning opt-out also imply that we should have the same for the removal from C++20? If that's the expectation, I would push back on that: we need the ability to make breaking changes (not huge ones) in less than 1 year.

**Comment by nico** [comment]
FWIW, a per-deprecation warning disable toggle for a single release would be more than enough for our use case, and it would make our lifes a lot easier.

**Comment by gulfemsavrun** [comment]
We started observing an interesting behavior in our codebase after this patch, and let me explain it with a small example below:

Case 1: There is no `-Wdeprecated-pragma` warning:
```
#include <errno.h>
int main() {
    return ENODATA;
}
```

Case2: There is a `-Wdeprecated-pragma` warning, where the only difference is `#include <iostream>`.
```
#include <errno.h>
#include <iostream>
int main() {
    return ENODATA;
}

test-pragma.cpp:5:12: warning: macro 'ENODATA' has been marked as deprecated: ENODATA is deprecated in ISO C++ [-Wdeprecated-pragma]
    5 |     return ENODATA;
      |            ^
/usr/local/google/home/gulfem/fuchsia/out/clang-0f6105/bin/../include/c++/v1/cerrno:42:71: note: macro marked 'deprecated' here
   42 | #  pragma clang deprecated(ENODATA, "ENODATA is deprecated in ISO C++")
      |                                                                       ^
```

Is this the expected behavior? `ENODATA` is defined in <errno.h> and is not deprecated. But, I think `include <iostream>` might transitively include some other headers like `<cerrno>` which causes `-Wdeprecated-pragma` warning.

**Comment by gulfemsavrun** [comment]
How should we proceed on this because we are getting a handful of these cases in our `third_party` code. If `Case 2` is a valid warning, we need to suppress this warning via `-Wno-deprecated-pragma` in our `third_party`. If it is an invalid warning, when can expect this to be fixed? I'm trying to find the best course of action at the moment because it is blocking our Clang roll.

**Comment by mordante** [comment]
Sorry this issue slipped under the radar.

The deprecation was an LWG issue that deprecates both the `errc` and `errno` macros.
https://cplusplus.github.io/LWG/issue3869. @AaronBallman (as former WG14 liason) do you know whether these changes have been proposed in WG14?

@gulfemsavrun would the toggle as suggested by @nico work for you?
Your example including `<iostream>` is indeed due to transitive includes.

**Comment by AaronBallman** [comment]
> The deprecation was an LWG issue that deprecates both the `errc` and `errno` macros. https://cplusplus.github.io/LWG/issue3869. @AaronBallman (as former WG14 liason) do you know whether these changes have been proposed in WG14?

Not former, I'm still the community liaison to WG14. :-) C doesn't actually standardize the deprecated macros in the first place, so there's nothing to propose there that I can see. These errno macros all came from POSIX, which did deprecate them (and was what spurred WG21 to deprecate them as well).

**Comment by mordante** [comment]
> > The deprecation was an LWG issue that deprecates both the `errc` and `errno` macros. https://cplusplus.github.io/LWG/issue3869. @AaronBallman (as former WG14 liason) do you know whether these changes have been proposed in WG14?
> 
> Not former, I'm still the community liaison to WG14. :-) 

My bad.

> C doesn't actually standardize the deprecated macros in the first place, so there's nothing to propose there that I can see. These errno macros all came from POSIX, which did deprecate them (and was what spurred WG21 to deprecate them as well).

Thanks for the information.

**Comment by AaronBallman** [comment]
> > > The deprecation was an LWG issue that deprecates both the `errc` and `errno` macros. https://cplusplus.github.io/LWG/issue3869. @AaronBallman (as former WG14 liason) do you know whether these changes have been proposed in WG14?
> > 
> > 
> > Not former, I'm still the community liaison to WG14. :-)
> 
> My bad.

No worries, that was more just so you and the rest of libc++ folks know that I'm happy to help with C-related stuff or going back to WG14 whenever you'd like.

**Comment by gulfemsavrun** [comment]
> @gulfemsavrun would the toggle as suggested by @nico work for you? Your example including `<iostream>` is indeed due to transitive includes.

We decided to globally suppress this warning to unblock our Clang roll. But, that's something temporary, and we would like to suppress this warning in fine-grained level. All of the cases that we currently see are like `Case 2`. That's why I would to like get a clarification whether `Case 2` is considered a valid warning before we invest into suppressing in fine-grained level. TBH, that does not seem right to see different warnings between `Case 1` and `Case 2`. Is there any plan to change that?

**Comment by AaronBallman** [comment]
I think the behavior is defensible per https://eel.is/c++draft/syserr#system.error.syn-1 (which I think is pulled in transitively via https://eel.is/c++draft/error.reporting), and also https://eel.is/c++draft/support.c.headers.general#1.sentence-3

(I'm not a libc++ developer though and they may have totally different opinions than me.)

**Comment by mordante** [comment]
@gulfemsavrun I like to discuss the way forward before deciding the changes I'm going to make.

I agree with @AaronBallman the behaviour is correct. However I wonder whether we're doing our users a service with the correct behaviour. When C++ removes the POSIX macros from `cerrno` in C++XX they may still be available in `errno.h`. In that case the user no longer gets a deprecation message from libc++ and can keep using them until their C library removes the macros. So some users may be happier when C++ removes them.

I think we should keep the behaviour for the `std::errc` and add a toggle for one release as requested by @nico.

For the macros I wonder what the best way would be:
1. remove the deprecation entirely
2. keep the deprecation, but only in C++26 (so don't retroactively apply this part of the LWG issue). Since there are few user of C++26 I see no strong motivation to add the `std::errc` toggle here.
3. keep the deprecation and use the `std::errc`  toggle for one release

Personally I'm leaning towards 2, but I like to hear your opinions.

**Comment by AaronBallman** [comment]
FWIW, the POSIX macros can be deprecated in `errno.h` as well, via `#pragma clang deprecate`: https://godbolt.org/z/f8hdcs3Mf

Personally, I think deprecation is only valuable if users are warned about it explicitly. So I tend to lean towards (3) because it gets the notification out to users. It might be worth reaching out to libc folks supporting POSIX to see if they'd consider adding the deprecation warnings to errno.h as well?

**Comment by ldionne** [comment]
Personally, I'd tend towards (1), i.e. do not deprecate the macros at all. After re-consideration, I feel like we might be overstepping the boundary between libc++ and the C Standard Library. We don't define these `errno.h` macros, the C Standard Library does. If they decide to mark these macros as deprecated as per POSIX, it's their choice. If they don't, it's their choice too and I feel like trying to force that upon them is only going to create compatibility problems between libc++ and the underlying C library.

@nico Were you asking for an opt-out for the `std::errc` constants being deprecated, or for the `<errno.h>` macros being deprecated? If the latter, we don't need to add one if we do (1). If you wanted an opt-out for the former, I think we can do that as a one-off thing but in general, deprecation warnings are introduced in newer standards only so that acts as the "opt in", as  Mark made me realize just now. So in general, I'd rather not start adding opt-outs for every single deprecation warning we add in newer standard modes. This `std::errc` deprecation is special though, since it is applied as a DR and so I think it would make sense to do a one-off opt-out.

**Comment by tbodt** [comment]
> At first blush, I was surprised by this. However, `ENODATA` is a POSIX errno value, and it is obsolescent in POSIX as well: https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/errno.h.html The same is true for `ETIME`, `ENOSTR`, and `ENOSR`.

The thing is, there's a name collision here. Linux never implemented STREAMS and thus never added these error codes as defined in POSIX, but they did define ENODATA as a Linux specific error used (afaik) only to mean "the extended attribute you tried to reference does not exist". This code is not deprecated and is often necessary in code that deals with extended attributes.

**Comment by ldionne** [comment]
Re above: So here again, simply not deprecating the macros in libc++ and letting the C Library handle that business seems to do the right thing. I think this reinforces that we should go for (1).

**Comment by mordante** [comment]
Based on the feedback and a private discussion with Louis I've implemented option 1 in #88296.

@gulfemsavrun That should disable the deprecation message in your `Case 1` and `Case 2`.
@nico From your messages I understand you only have an issue with the deprecation of the macros. The new PR should fix your issues too. This means there is no need for a deprecation toggle anymore. Please let me know when you still need a deprecation toggle.

**Comment by gulfemsavrun** [comment]
> @gulfemsavrun That should disable the deprecation message in your `Case 1` and `Case 2`.

Thanks for the update @mordante!

**Comment by AaronBallman** [comment]
FWIW, I'm also happy enough going with option 1, so thank you for the quick fix here!

**Comment by nico** [comment]
> Were you asking for an opt-out for the std::errc constants being deprecated, or for the <errno.h> macros being deprecated? If the latter, we don't need to add one if we do (1)
> From your messages I understand you only have an issue with the deprecation of the macros.

Yes, correct, just the errno.h macros. Thanks!
